### PR TITLE
feat(bindings): Add logo_uri to OidcConfiguration.

### PR DESCRIPTION
**Pull request by pixlwave**
Part of https://github.com/vector-im/element-x-ios/issues/1547 this PR updates the `OidcConfiguration` to take a logo URL that will be displayed in the Consent prompt.

Whilst this property is optional in the request, without it MAS displays a big white square so I chose to not make it an optional property on the configuration.

### CODE REVIEW

**Comment by jplatte** [comment]
> Whilst this property is optional in the request, without it MAS displays a big white square so I chose to not make it an optional property on the configuration.

Can you please make it optional? I don't think the current implementation of MAS should determine what the API looks like.

File: bindings/matrix-sdk-ffi/src/authentication_service.rs [hunk 1]
```diff
@@ -109,6 +109,8 @@ pub struct OidcConfiguration {
     pub redirect_uri: String,
     /// A URI that contains information about the client.
     pub client_uri: String,
+    /// A URI that contains the client's logo.
+    pub logo_uri: String,
```
Comment (Hywan) [inline bindings/matrix-sdk-ffi/src/authentication_service.rs:113]:
```suggestion
    pub logo_uri: Option<String>,
```

File: bindings/matrix-sdk-ffi/src/authentication_service.rs [hunk 2]
```diff
@@ -556,6 +558,7 @@ impl AuthenticationService {
             .map_err(|_| AuthenticationError::OidcCallbackUrlInvalid)?;
         let client_name = Some(Localized::new(configuration.client_name.to_owned(), []));
         let client_uri = Url::parse(&configuration.client_uri).ok().map(|u| Localized::new(u, []));
+        let logo_uri = Url::parse(&configuration.logo_uri).ok().map(|u| Localized::new(u, []));
```
Comment (Hywan) [inline bindings/matrix-sdk-ffi/src/authentication_service.rs:561]:
```suggestion
        let logo_uri = configuration
            .logo_uri
            .as_deref()
            .map(Url::parse)
            .transpose()
            .map_err(|_| AuthenticationError::OidcMetadataInvalid)?
            .map(|u| Localized::new(u, []));
```

I'm not sure about the error to return though.

Comment (pixlwave) [reply bindings/matrix-sdk-ffi/src/authentication_service.rs:561]: That error looks correct to me 👍

**Comment by pixlwave** [comment]
Updated, all the properties are now optional apart from redirect_uri.

File: bindings/matrix-sdk-ffi/src/authentication_service.rs [hunk 3]
```diff
@@ -556,11 +556,36 @@ impl AuthenticationService {
...
+            .transpose()
+            .map_err(|_| AuthenticationError::OidcMetadataInvalid)?
+            .map(|u| Localized::new(u, []));
+        let logo_uri = configuration
+            .logo_uri
+            .as_deref()
+            .map(Url::parse)
+            .transpose()
+            .map_err(|_| AuthenticationError::OidcMetadataInvalid)?
+            .map(|u| Localized::new(u, []));
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/authentication_service.rs:574]: Might be easier to read?

```suggestion
            .map(|uri| {
                Ok(Localized::new(
                    Url::parse(uri).map_err(|_| AuthenticationError::OidcMetadataInvalid)?,
                    [],
                ))
            })
            .transpose()?;
```

Comment (pixlwave) [reply bindings/matrix-sdk-ffi/src/authentication_service.rs]: Updated.

Comment (pixlwave) [reply bindings/matrix-sdk-ffi/src/authentication_service.rs]: Also, I tried doing an impl on `Option<&str>` to reduce duplicated code, but I don't think that's allowed in Rust?

Comment (pixlwave) [reply bindings/matrix-sdk-ffi/src/authentication_service.rs]: Updated with an extension trait 😃

**Comment by pixlwave** [comment]
I'm not sure I follow how the changes I've made have resulted in those compile errors 🤔

**Comment by jplatte** [comment]
Unrelated, new clippy lints from the latest Rust version that was released today.

**Review by jplatte — APPROVED** [review]

**Pull request merged**